### PR TITLE
fixed float compare to 0 for fast-math

### PR DESCRIPTION
**Pull request by chacha21**
Bug #24019

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [X] There is a reference to the original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by chacha21** [comment]
I have checked that `countNonZero()` has the same original problem.
Under fast-math, without IPP, it does not count `NaN `as` != 0`.
So, should `countNonZero(`) be fixed as well, or let `hasNonZero()` behave the same ?

**Comment by chacha21** [comment]
> nz += (x+x) != 0;

Any reason not to write "x<<1" ?

File: modules/core/test/test_hasnonzero.cpp [hunk 1]
```diff
@@ -81,6 +81,26 @@ INSTANTIATE_TEST_CASE_P(Core, HasNonZeroNegZeros,
...
+
+TEST_P(HasNonZeroNaNValues, hasNonZeroNaNValues)
+{
+    const int type = std::get<0>(GetParam());
+    const Size size = std::get<1>(GetParam());
+
+    Mat m = Mat(size, type);
+
+    m.setTo(Scalar::all(std::numeric_limits<double>::quiet_NaN()));
+    EXPECT_TRUE(hasNonZero(m));
```
[the remark this answers is not in the document]
Comment (chacha21) [reply modules/core/test/test_hasnonzero.cpp:94]: > I have checked that countNonZero() has the same original problem (...) So, should countNonZero() be fixed as well, or let hasNonZero() behave the same ?

That was my question, I see several answers :

1. stick to the original behaviour (do not care about fast-math NaN problems)
2. fix every function to get a new behaviour (handling NaN correctly under fast-math) (ouch)
3. "tag" functions to explicitely document whether they use "fast" or "correct" fast-math NaN as an opt-in behaviour

A preprocess macro inherited from CMake could help by publishing "is fast math enabled" information at compile time.

**Comment by vpisarev** [comment]
@chacha21, the solution that I proposed earlier `(x + x) != 0` or `(x << 1) != 0`, where 'x' is integer representation of the floating-point number, has been implemented in 5.x branch: https://github.com/opencv/opencv/blob/5.x/modules/core/src/has_non_zero.simd.hpp. If this functionality is still needed in 4.x, I think, it makes sense to backport that code. Let me close this PR that introduces unnecessary slowdown

**Pull request closed without merging**